pkg/virt-controller/watch/snapshot.go

```diff
@@ -0,0 +1,777 @@
+var currentTime = func() *metav1.Time {
+	t := metav1.Now()
+	return &t
+}
+
+func (ctrl *SnapshotController) updateVMSnapshot(vmSnapshot *vmssv1alpha1.VirtualMachineSnapshot) error {
+	log.Log.V(3).Infof("Updating VirtualMachineSnapshot %s/%s", vmSnapshot.Namespace, vmSnapshot.Name)
+
+	// Make sure status is initialized
+	if vmSnapshot.Status == nil {
+		return ctrl.updateSnapshotStatus(vmSnapshot)
+	}
+
+	source, err := ctrl.getSnapshotSource(vmSnapshot)
+	if err != nil {
+		return err
+	}
+
+	// unlock the source if done/error
+	if !vmSnapshotProgressing(vmSnapshot) && source != nil && source.Locked() {
+		if err = source.Unlock(); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// check deleted
+	if vmSnapshot.DeletionTimestamp != nil {
+		return ctrl.cleanupVMSnapshot(vmSnapshot)
+	}
+
+	if source != nil && vmSnapshotProgressing(vmSnapshot) {
+		// attempt to lock source
+		// if fails will attempt again when source is updated
+		if !source.Locked() {
+			locked, err := source.Lock()
+			if err != nil {
+				return err
+			}
+
+			log.Log.V(3).Infof("Attempt to lock source returned: %t", locked)
+
+			return nil
+		}
+
+		// add source finalizer and maybe other stuff
+		updated, err := ctrl.initVMSnapshot(vmSnapshot)
+		if updated || err != nil {
+			return err
+		}
+
+		content, err := ctrl.getContent(vmSnapshot)
+		if err != nil {
+			return err
+		}
+
+		// create content if does not exist
+		if content == nil {
+			return ctrl.createContent(vmSnapshot)
+		}
+	}
+
+	if err = ctrl.updateSnapshotStatus(vmSnapshot); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ctrl *SnapshotController) updateVMSnapshotContent(content *vmssv1alpha1.VirtualMachineSnapshotContent) error {
+	log.Log.V(3).Infof("Updating VirtualMachineSnapshotContent %s/%s", content.Namespace, content.Name)
+
+	var volueSnapshotStatus []vmssv1alpha1.VolumeSnapshotStatus
+	var deletedSnapshots, skippedSnapshots []string
+
+	currentlyReady := content.Status != nil && content.Status.ReadyToUse != nil && *content.Status.ReadyToUse
+	currentlyError := content.Status != nil && content.Status.Error != nil
+
+	for _, volumeBackup := range content.Spec.VolumeBackups {
+		if volumeBackup.VolumeSnapshotName == nil {
+			continue
+		}
+
+		vsName := *volumeBackup.VolumeSnapshotName
+
+		volumeSnapshot, err := ctrl.getVolumeSnapshot(content.Namespace, vsName)
+		if err != nil {
+			return err
+		}
+
+		if volumeSnapshot == nil {
+			// check if snapshot was deleted
+			if currentlyReady {
+				log.Log.Warningf("VolumeSnapshot %s no longer exists", vsName)
+				ctrl.recorder.Eventf(
+					content,
+					corev1.EventTypeWarning,
+					volumeSnapshotMissingEvent,
+					"VolumeSnapshot %s no longer exists",
+					vsName,
+				)
+				deletedSnapshots = append(deletedSnapshots, vsName)
+				continue
+			}
+
+			if currentlyError {
+				log.Log.V(3).Infof("Not creating snapshot %s because in error state", vsName)
+				skippedSnapshots = append(skippedSnapshots, vsName)
```

Comment:
When an error occurs, does that mean the snapshot will never recover? It appears that way since we no longer process volume snapshots once an error occurs.

Reply:
For the most part, yes.  Would rather fail an operation than keep sources locked until success.

The one recoverable error is when a VolumSnapshot disappears and then reappears (without us recreating it).  This may happen if a storage administrator mistakenly deletes/restores a VolumeSnapshot.  Perhaps an unlikely scenario but something worth supporting I think.

Reply:
yep, understood